In [21]:
#Cargamos librerias
import pickle
from sklearn.ensemble import BaggingClassifier
from sklearn.model_selection import cross_val_score
import numpy as np
import xgboost as xgb
from sklearn.metrics import classification_report
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier





In [22]:
# Cargar los datos de entrenamiento muestreados desde el archivo pickle
with open(r'C:\Users\ocata\OneDrive\Desktop\Proyecto7_DS\data\processed\datos_entrenamiento_muestreados.pkl', 'rb') as f:
    X_train_resampled, y_train_resampled = pickle.load(f)

# Cargar el modelo entrenado desde el archivo pickle
with open(R'C:\Users\ocata\OneDrive\Desktop\Proyecto7_DS\models\mejor_rf_modelo.pkl', 'rb') as f:
    best_rf_classifier = pickle.load(f)

# Cargar las variables desde el archivo pickle
with open(R'C:\Users\ocata\OneDrive\Desktop\Proyecto7_DS\data\processed\datos_split_procesados.pkl', 'rb') as f:
    X_train_pca, X_test_pca, y_train, y_test = pickle.load(f)


# Utilizamos la tecnica de BaggingClassifer

In [23]:

# Número de modelos a crear (puedes ajustar este valor según tus necesidades)
num_modelos = 27

# Crear un clasificador de Bagging con el modelo RandomForestClassifier como estimador base
bagging_classifier = BaggingClassifier(estimator=best_rf_classifier, 
                                       n_estimators=num_modelos, 
                                       random_state=42)

# Entrenar el clasificador de Bagging
bagging_classifier.fit(X_train_resampled, y_train_resampled)

# Calcular la precisión del clasificador de Bagging en el conjunto de prueba
bagging_accuracy = bagging_classifier.score(X_test_pca, y_test)
print("Precisión del clasificador de Bagging:", bagging_accuracy)




Precisión del clasificador de Bagging: 0.6382978723404256


# Utillizamos la tecnica de AdaBoostClassifier

In [24]:

# Crear un clasificador de AdaBoost con el modelo RandomForestClassifier como base
ada_classifier = AdaBoostClassifier(estimator=best_rf_classifier, n_estimators=50, random_state=42)

# Entrenar el clasificador de AdaBoost
ada_classifier.fit(X_train_resampled, y_train_resampled)

# Realizar predicciones en el conjunto de prueba
y_pred_ada = ada_classifier.predict(X_test_pca)

# Calcular la precisión del clasificador de AdaBoost
ada_accuracy = ada_classifier.score(X_test_pca, y_test)
print("Precisión del clasificador de AdaBoost:", ada_accuracy)

# Mostrar el informe de clasificación
print("Informe de Clasificación (AdaBoost):\n", classification_report(y_test, y_pred_ada))


Precisión del clasificador de AdaBoost: 0.6595744680851063


c:\Users\ocata\miniconda3\envs\proyecto7\Lib\site-packages\sklearn\ensemble\_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
c:\Users\ocata\miniconda3\envs\proyecto7\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\ocata\miniconda3\envs\proyecto7\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Informe de Clasificación (AdaBoost):
               precision    recall  f1-score   support

           0       0.00      0.00      0.00         4
           1       0.69      0.91      0.78        32
           2       0.40      0.18      0.25        11

    accuracy                           0.66        47
   macro avg       0.36      0.36      0.34        47
weighted avg       0.56      0.66      0.59        47



c:\Users\ocata\miniconda3\envs\proyecto7\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


# Utilizamos la tecnica de GradientBoostingClassifier

In [25]:

# Crear un clasificador de Gradient Boosting
gb_classifier = GradientBoostingClassifier(n_estimators=100, random_state=42)

# Entrenar el clasificador de Gradient Boosting
gb_classifier.fit(X_train_resampled, y_train_resampled)

# Realizar predicciones en el conjunto de prueba
y_pred_gb = gb_classifier.predict(X_test_pca)

# Calcular la precisión del clasificador de Gradient Boosting
gb_accuracy = gb_classifier.score(X_test_pca, y_test)
print("Precisión del clasificador de Gradient Boosting:", gb_accuracy)

# Mostrar el informe de clasificación
print("Informe de Clasificación (Gradient Boosting):\n", classification_report(y_test, y_pred_gb))


Precisión del clasificador de Gradient Boosting: 0.48936170212765956
Informe de Clasificación (Gradient Boosting):
               precision    recall  f1-score   support

           0       1.00      0.25      0.40         4
           1       0.63      0.59      0.61        32
           2       0.19      0.27      0.22        11

    accuracy                           0.49        47
   macro avg       0.61      0.37      0.41        47
weighted avg       0.56      0.49      0.50        47



# Utilizamos la tecnica de XGBClassifier

In [26]:
# Crear un clasificador de XGBoost
xgb_classifier = xgb.XGBClassifier(n_estimators=100, random_state=42)

# Entrenar el clasificador de XGBoost
xgb_classifier.fit(X_train_resampled, y_train_resampled)

# Realizar predicciones en el conjunto de prueba
y_pred_xgb = xgb_classifier.predict(X_test_pca)

# Calcular la precisión del clasificador de XGBoost
xgb_accuracy = xgb_classifier.score(X_test_pca, y_test)
print("Precisión del clasificador de XGBoost:", xgb_accuracy)

# Mostrar el informe de clasificación
print("Informe de Clasificación (XGBoost):\n", classification_report(y_test, y_pred_xgb))


Precisión del clasificador de XGBoost: 0.6382978723404256
Informe de Clasificación (XGBoost):
               precision    recall  f1-score   support

           0       0.00      0.00      0.00         4
           1       0.72      0.81      0.76        32
           2       0.36      0.36      0.36        11

    accuracy                           0.64        47
   macro avg       0.36      0.39      0.38        47
weighted avg       0.58      0.64      0.61        47



c:\Users\ocata\miniconda3\envs\proyecto7\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\ocata\miniconda3\envs\proyecto7\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\ocata\miniconda3\envs\proyecto7\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(r

# Realizamos una validacion cruzada

In [27]:
# Validación cruzada con el modelo Random Forest original
cv_scores = cross_val_score(best_rf_classifier, X_train_resampled, y_train_resampled, cv=5)
print("Puntuaciones de validación cruzada:", cv_scores)
print("Precisión media de validación cruzada:", np.mean(cv_scores))


Puntuaciones de validación cruzada: [0.75384615 0.84375    0.78125    0.875      0.859375  ]
Precisión media de validación cruzada: 0.8226442307692308


# Conclusiones


Ninguna de las tecnicas de emsabling arrojó mejores resultados para el modelo. Por otro lado al realizar una validación cruzada tenemos un puntaje del 82% lo cual da cuenta de un modelo de bastante robustes, quedandonos con este modelo para producción.